CMake unity builds; core_tests as an example

mj-xmr opened the pull request:
core_tests build 50% faster. 

description follows soon.

Work time: 2h

mj-xmr commented on CMakeLists.txt:
```diff
@@ -75,7 +75,23 @@ if (${CMAKE_VERSION} VERSION_GREATER "3.0.0" AND CMAKE_MAKE_PROGRAM MATCHES "nin
 endif ()
-  
+
+option (USE_UNITY "Use unity builds" ON) # TODO: switch to OFF before review
```
I wanted to test if the produced binaries deliver good results on your CI.

mj-xmr replied:
All the targets, which build core-tests, have succeeded in building the target via unity build. I'll leave this until you review and decide if we should maybe keep in ON by default.

xiphon commented:
Should break anonymous namespaces, i think.

mj-xmr commented:
> Should break anonymous namespaces, i think.

I've just checked. It breaks **equally named** variables in anonymous namespaces in the same way as it breaks static global variables.

mj-xmr commented:
Figured out how to do a Unity Build for unit tests for a 36% gain.

jeffro256 commented:
Are the build commands different this far back? I can't get this branch to build. What do you recommend I run? (My sub modules are synced for this commit btw)

mj-xmr commented:
Excluded `unit_tests` from this PR and placed them under [this branch](https://github.com/mj-xmr/monero-mj/tree/cmake-unity-build-unit-tests), as UTs require some more work than Core Tests.

jeffro256 commented:
Results from building 98e4e96:

With `USE_UNITY=OFF`:

    real	49m37.932s
    user	47m25.331s
    sys	2m10.006s

With `USE_UNITY=ON`:

    real	48m44.421s
    user	46m33.999s
    sys	2m8.321s

mj-xmr commented:
> I don't see any glaring issues with this PR as long as there are no name conflicts, which might be easier said than done.

I can tell you, that for the `core_tests` it's much easier to avoid name conflicts, than it is for `unit_tests`, which evolve faster (which is good), but at the same time leave much more mess behind as a by-product.

> Would you plan on eventually making each sub directory of `src/` its own unity build?

It seems that it only makes sense to do it for the core- and unit tests, since they are clearly 2 groups of very similar sources. While it's possible to enable UBuilds for other sub-directories, at the current situation, where they aren't as spread out as the mentioned tests, this would not bring much benefit.

However, I could imagine, that some time in the future, when maybe `wallet7` becomes nicely concept-separated, where each concept is put into a separate class and file, there might be a need to join the similar pieces automatically by the UBuild. Having a tool for this at hand means that we're prepared for such a situation.

jeffro256 commented:
> However, I could imagine, that at point in the future, when maybe `wallet7` becomes nicely concept-separated, where each concept is put into a separate class and file, there might be a need to join the similar pieces automatically by the UBuild

Don't hold your breath ;)

mj-xmr commented:
@jeffro256 Thanks for approving. As I wrote initially, it would probably be better if the option was OFF by default, as it can be confusing and is unproductive when actually writing individual tests, expecting the incremental compilation to be as fast as possible. Also we have this rule here, that except for `ccache`, which is an industry standard, we keep all such options OFF by default.

So as soon as The Management, ehm... @selsta is OK with merging this one, I will change the option to OFF.

jeffro256 commented:
Yeah it's a really cool feature, and should be documented so people don't miss it, but I can imagine it could cause weird, hard to debug issues if it was ON by default.

mj-xmr commented:
> Yeah it's a really cool feature, and should be documented so people don't miss it (...)

There's already a hook for this [in the documentation](https://github.com/monero-project/monero/blob/master/docs/COMPILING_DEBUGGING_TESTING.md#unity-builds) that I had prepared, as well as for other such PRs.

mj-xmr commented:
So, @selsta : ready to merge as soon as I switch it to OFF by default?

@jeffro256 : I advice to tick all the files as "viewed" so that you only have to recheck the single CMakeLists.txt that will change.

selsta commented:
`-j8, WITH_UNITY=ON`

```
Executed in  129.63 secs    fish           external
   usr time  715.76 secs    0.14 millis  715.76 secs
   sys time   50.75 secs    1.36 millis   50.75 secs
```

`-j8, WITH_UNITY=OFF`
```
Executed in  129.84 secs    fish           external
   usr time  715.30 secs    0.14 millis  715.30 secs
   sys time   50.24 secs    1.61 millis   50.24 secs
```

`-j1, WITH_UNITY=ON`
```
Executed in  701.12 secs    fish           external
   usr time  654.47 secs    0.15 millis  654.47 secs
   sys time   40.47 secs    1.61 millis   40.47 secs
```

`-j1, WITH_UNITY=OFF`
```
Executed in  702.87 secs    fish           external
   usr time  654.72 secs    0.16 millis  654.72 secs
   sys time   40.45 secs    3.27 millis   40.45 secs
```

So far I'm not convinced at the usefulness of this. This is something that has to be enabled manually (enable by default has its own set of downsides) and the speedup is minimal in our case.

selsta commented:
`-j8, WITH_UNITY=ON`

```
Executed in  129.63 secs    fish           external
   usr time  715.76 secs    0.14 millis  715.76 secs
   sys time   50.75 secs    1.36 millis   50.75 secs
```

`-j8, WITH_UNITY=OFF`
```
Executed in  129.84 secs    fish           external
   usr time  715.30 secs    0.14 millis  715.30 secs
   sys time   50.24 secs    1.61 millis   50.24 secs
```

`-j1, WITH_UNITY=ON`
```
Executed in  701.12 secs    fish           external
   usr time  654.47 secs    0.15 millis  654.47 secs
   sys time   40.47 secs    1.61 millis   40.47 secs
```

`-j1, WITH_UNITY=OFF`
```
Executed in  702.87 secs    fish           external
   usr time  654.72 secs    0.16 millis  654.72 secs
   sys time   40.45 secs    3.27 millis   40.45 secs
```

So far I'm not convinced at the usefulness of this. This is something that has to be enabled manually (enable by default has its own set of downsides) and the speedup is minimal in our case.

mj-xmr commented:
Default OFF is a sure thing.

I'll re-run my benchmark and provide a script, that we can all reuse.

selsta commented:
- RAM disk,
- `cmake -DUSE_CCACHE=OFF -DBUILD_TESTS=ON -DBUILD_SHARED_LIBS=ON -DSTRIP_TARGETS=ON`
- `time make core_tests`

`-DWITH_UNITY=ON`
```
Executed in  357.49 secs    fish           external
   usr time  343.59 secs    0.15 millis  343.59 secs
   sys time   12.73 secs    3.23 millis   12.73 secs
```

`-DWITH_UNITY=OFF`
```
Executed in  410.79 secs    fish           external
   usr time  394.05 secs    0.17 millis  394.05 secs
   sys time   15.42 secs    1.86 millis   15.42 secs
```

Difference: 13%

-----------------------

- SSD,
- `cmake -DUSE_CCACHE=OFF -DBUILD_TESTS=ON`
- `time make core_tests`

`-DWITH_UNITY=ON`
```
Executed in  356.77 secs    fish           external
   usr time  342.52 secs    0.25 millis  342.52 secs
   sys time   12.23 secs    5.03 millis   12.22 secs
```

`-DWITH_UNITY=OFF`
```
Executed in  400.95 secs    fish           external
   usr time  386.12 secs    0.15 millis  386.12 secs
   sys time   13.23 secs    1.74 millis   13.23 secs
```

Difference: 11%

mj-xmr commented:
Rebased and added a throw away commit, that tests the build time objectively.

I invite everybody to try for themselves:

```git clone --recursive https://github.com/mj-xmr/monero-patches.git```
```time ./monero-patches/src/monero-test-build-time.sh cmake-unity-build tests/core_tests```

It will cost you however about 8 hours.

mj-xmr commented:
I'm not sure if you receive notifications about changes in the description of the PR. If yes, then I apologize for the spam.

What's changed:
- I created a [build time script](https://github.com/mj-xmr/monero-patches/blob/master/src/monero-test-build-time.sh) for you to use, in order to achieve results closer to a developer's reality
- I created and integrated [an automated CI Job](https://github.com/mj-xmr/monero-mj/runs/6719989492?check_suite_focus=true)  that performs the build time tests objectively.
- I updated the PR's description and updated the measurements:

measurement of |  master | this branch | reduction
 --- | ---  | --- | ---
complete build |         44.43 m |  41.35 m | -6.93%
the most affected target |         5.3 m |   2.833 m | -46.55%

These measurements come straight from the data available via the GH Action, so there's no risk of data manipulation from my side.
